c10/core/Scalar.h

```diff
@@ -26,6 +27,49 @@ class C10_API Scalar {
  public:
   Scalar() : Scalar(int64_t(0)) {}
 
+  Scalar(const Scalar& rhs) : tag(rhs.tag), v(rhs.v) {
+      if (isComplex()) {
+        c10::raw::intrusive_ptr::incref(v.z);
+      }
+  }
+
+  Scalar(Scalar&& rhs) noexcept : tag(rhs.tag), v(rhs.v) {}
```

Comment:
this is likely a bug causing a double free. After this, you now have two Scalar objects that both are going to destroy the same complex. Could be fixed by setting the *this scalar to not store a complex anymore similar to what you're doing in the move assignment operator

Reply:
as a side note: Having all four copy/move constructors and assignment operators implemented explicitly is some code duplication with the chance of bugs like this one. There's a nice pattern where you can implement all four of those in terms of just one implementation of `swap`, and then you have all the refcounting or destroying of state in just one function.

Reply:
Nice catch! The RHS complex will be cleaned up in move assignment but forget to to the same cleanup in move constructor.

Reply:
@smessmer : Nice catch! The RHS complex will be cleaned up in move assignment but forget to to the same cleanup in move constructor.